offer-manager-domain/src/main/groovy/life/qbic/business/offers/Offer.groovy

```diff
@@ -478,24 +485,29 @@ class Offer {
         return calculateTotalDiscountAmount()
     }
 
-    private double calculateTotalDiscountAmount() {
-        return items.sum{ it.quantityDiscount }
+    private BigDecimal calculateTotalDiscountAmount() {
+        return items.sum{ it.quantityDiscount } as BigDecimal
+    }
+
+    private BigDecimal calculateCataloguePriceForItem(ProductItem item) {
+        return item.quantity * determineUnitPrice(item) as BigDecimal
     }
 
     private BigDecimal discountAmountForProductItem(ProductItem productItem) {
         BigDecimal discount = 0
         if (productItem.product instanceof PrimaryAnalysis
                 || productItem.product instanceof SecondaryAnalysis) {
-            discount = quantityDiscount.apply(productItem.quantity as Integer, calculateItemNet(productItem))
+            BigDecimal itemCost = productItem.quantity * determineUnitPrice(productItem)
+            discount = quantityDiscount.apply(productItem.quantity as Integer, itemCost)
         }
         return discount
     }
 
-    private double calculateNetPrice() {
+    private BigDecimal calculateNetPrice() {
         if (items.empty) {
             return 0
         }
-        return items.sum {calculateItemNet(it)} as double
+        return items.sum {it.totalPrice - it.quantityDiscount} as BigDecimal
```

Comment:
I do not know how the casting to BigDecimal is handled by groovy.
In case this calls new BigDecimal(yourDoubleValue) this is not the desired behaviour (as stated by the BigDecimal documentation).
You should use `new BigDecimal(double.toString())` or `BigDecimal.valueOf(double)`